sdk/cxx/include/sawtooth_sdk.h

```diff
@@ -0,0 +1,258 @@
+    // of TransactionApplicator that will handle the processing. As little work
+    // as possible should be done in this function.
+    virtual TransactionApplicatorUPtr GetApplicator(
+        TransactionUPtr txn,
+        GlobalStateUPtr state) = 0;
+};
+typedef std::unique_ptr<TransactionHandler> TransactionHandlerUPtr;
+typedef std::shared_ptr<TransactionHandler> TransactionHandlerPtr;
+
+
+// Helper class to provide mappings from domain identifiers
+// to merkle trie addresses. This class is designed to be overloaded, such
+// that implementers of TransactionProcessors can define their own mappings
+// from their native object ids to merkle trie addresses.
+//
+// The default implementation uses a SHA512 hash to map both the key and the
+// namespace to the merkle trie address. Collisions in the key mappings should
+// be uncommon but still possible, data storage behind the keys.
+//
+// Each instance of the mapper holds a namespace that is mapped to a
+// namespace prefix when it is constructed. If your transaction processor
+// operates on multiple namespaces, it is recommended that you use an instance
+// for each namespace.
+class AddressMapperIF {
+ public:
+
+    virtual ~AddressMapperIF() {}
+
+    // Maps an namespace string to an namespace address prefix.
+    // this provides a default implementation using SHA512.
+    // it is intended to be overridden by subclass implementations
+    // that can provide their own mapping of namespace strings to
+    // prefixes. The requirements are that the mapping needs to
+    // return a prefix that is 6 characters in length and consists
+    // of only lower case hexadecimal characters(0123456789abcdef).
+    virtual std::string MapNamespace(const std::string& key) const = 0;
+
+    // Maps an key string to a address, this provides a default
+    // implementation using SHA512.
+    // It is intended to be overridden by subclass implementations
+    // that can provide their own mappings. The requirements are that
+    // the mapping needs to return a string that is 128 characters
+    // in length and consists of only lower case hexadecimal characters.
+    virtual std::string MapKey(const std::string& key) const = 0;
+
+
+    // returns the namespace prefix generated when the class was constructed.
+    virtual std::string GetNamespacePrefix() = 0;
+
+    // Maps the key passed to the function and concatenates it with the
+    // namespace prefix this object was initialized with and returns
+    // a fully qualified Merkle Trie address. These addresses are always
+    // 134 characters in length and consist entirely of lower case hexadecimal
+    // characters(0123456789abcdef).
+    virtual std::string MakeAddress(const std::string& key) = 0;
+
+
+    static AddressMapperIF* Create(const std::string& namespace_);
+};
+
+typedef std::shared_ptr<AddressMapperIF> AddressMapperPtr;
+typedef std::unique_ptr<AddressMapperIF> AddressMapperUPtr;
+
+
+
+class TransactionProcessorIF
```

Comment:
There is only one implementation fo the TransactionProcessor, so the interface does not seem necessary.

Reply:
The primary goal of this interface is to isolate app developers from underlying details of TransactionProcessor implementation so no SDK code is required for building a TP.